### PR TITLE
PARQUET-1827: UUID type currently not supported by parquet-mr

### PR DESCRIPTION
**Pull request by gszadovszky**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java [hunk 1]
```diff
@@ -861,6 +871,36 @@ PrimitiveStringifier valueStringifier(PrimitiveType primitiveType) {
     }
   }
 
+  public static class UUIDLogicalTypeAnnotation extends LogicalTypeAnnotation {
```
Comment (shangxinli) [inline parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java:874]: Why there is no implementation for hasCode() and equal()?

File: parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java [hunk 2]
```diff
@@ -421,4 +422,30 @@ private String stringifyWithScale(BigInteger i) {
       }
     };
   }
+
+  static final PrimitiveStringifier UUID_STRINGIFIER = new PrimitiveStringifier("UUID_STRINGIFIER") {
+    private final char[] digit = "0123456789abcdef".toCharArray();
+    @Override
+    public String stringify(Binary value) {
+      byte[] bytes = value.getBytesUnsafe();
```
Comment (shangxinli) [inline parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java:430]: Do you want to make 'bytes' final since you are calling getBytesUnsafe()?

File: parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java [hunk 3]
```diff
@@ -309,6 +308,35 @@ public void testDecimalStringifier() {
...
+    PrimitiveStringifier stringifier = PrimitiveStringifier.UUID_STRINGIFIER;
+
+    assertEquals("00112233-4455-6677-8899-aabbccddeeff", stringifier.stringify(
+        toBinary(0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff)));
+    assertEquals("00000000-0000-0000-0000-000000000000", stringifier.stringify(
+        toBinary(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)));
+    assertEquals("ffffffff-ffff-ffff-ffff-ffffffffffff", stringifier.stringify(
+        toBinary(0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff)));
+
+    assertEquals("0eb1497c-19b6-42bc-b028-b4b612bed141", stringifier.stringify(
```
Comment (shangxinli) [inline parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java:322]: Are the 3 test data are randomly chosen? It seems duplicate coverage. 

Could you add some negative tests like incorrect length uuids, invalid characters etc.

File: parquet-avro/README.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<!--
```
Comment (shangxinli) [inline parquet-avro/README.md:1]: Should we add it using separate jira?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java [hunk 5]
```diff
@@ -766,6 +768,33 @@ public void testReuseNameInNestedStructureAtSameLevel() throws Exception {
...
+        Arrays.asList(new Schema.Field("uuid", LogicalTypes.uuid().addToSchema(Schema.create(STRING)), null, null)));
+    String parquet = "message myrecord {\n" +
+        "  required binary uuid (STRING);\n" +
+        "}\n";
+    Schema toAvro = Schema.createRecord("myrecord", null, null, false,
+        Arrays.asList(new Schema.Field("uuid", Schema.create(STRING), null, null)));
+
+    testAvroToParquetConversion(fromAvro, parquet);
+    testParquetToAvroConversion(toAvro, parquet);
+  }
```
Comment (shangxinli) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: Can we have checkReaderWriterCompatibility() to verify if the parquet and avro schema are compatible for UUID?  
There are issues like PARQUET-1681 for avro schema and parquet schema conversion for other types.

File: parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java (hunk 1 above)
Comment (gszadovszky) [reply parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java:874]: As this is a singleton the default implementation of `equals(Object)` and `hashCode()` fits perfectly.

File: parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java (hunk 2 above)
Comment (gszadovszky) [reply parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java:430]: `final` would protect the reference only and not the values of the array. Making a reference final in a local scope is usually required in situations where it is accessed from e.g. lambda closures.

File: parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java (hunk 3 above)
Comment (gszadovszky) [reply parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java:322]: The idea is to have 3 kind-of corner cases and 3 common (random but constant) values. What do you mean by duplicate coverage? (I think, we do not need exhaustive testing for the stringifiers because they only used by our tools for debugging purposes.)

The stringifiers do not validate the data they get because of performance reasons. So, if the array is longer than 16 it would simply stringify the first 16 and skip the others. In case of the length is too short then an `ArrayIndexOutOfBoundsException` would be thrown. Do you think we should test these cases? They would not reach any additional branches in the parquet code.
Invalid characters are not possible. The full set of values of the 16 bytes array is covered in UUID.

File: parquet-avro/README.md (hunk 4 above)
Comment (gszadovszky) [reply parquet-avro/README.md:1]: Usually we separate jiras in similar cases to not make cherry-picking hard in case of the related changed needs to be on another branch as well. In this case this is only a documentation so should not cause any troubles. It would be cleaner if this documentation would have already been existed and I've had to add the docs of the new keys only (which would clearly be part of this change).
If you have a strong opinion to separate this to another change I'm happy to do so, though.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 5 above)
Comment (gszadovszky) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: To be honest I am not too familiar with parquet-avro. I've made the changes based on the implementation/test of other logical types. Could you explain it in more details what you would test exactly?

Comment (shangxinli) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: Basically we found some type of avro schema is not compatible with the parquet schema which it is converted to. This caused problem that the data cannot be read. I have a test here (https://github.com/shangxinli/parquet-mr/commit/f80469f55b83404ea334ee4019f658ecdb5ac575#diff-536ca67880a7870cf8df8f95143bd7d7R814) that reproduce the issue for a nested schema. I know likely UUID type won't have this issue but I better to have a test for it. It is pretty easy to add also.

File: parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java (hunk 3 above)
Comment (shangxinli) [reply parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java:322]: By duplicate coverage, I meant the #2, #3 tests seems repeating the same test as #1. The value is different, but when the test executes, they would execute the same code path. So I think they won't provide extra coverage. 

From test perspective, negative test does provide values. In this case, we can test the exception is thrown as expected if it is too short. 

For "if the array is longer than 16 it would simply stringify the first 16 and skip the others", that could cause silent errors, right?

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 5 above)
Comment (shangxinli) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: If it is too much effort for doing this, it is OK not to do it. It is a lower priority.

File: parquet-avro/README.md (hunk 4 above)
Comment (shangxinli) [reply parquet-avro/README.md:1]: Sounds good

File: parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java (hunk 2 above)
Comment (shangxinli) [reply parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveStringifier.java:430]: Sounds good!

File: parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java (hunk 1 above)
Comment (shangxinli) [reply parquet-column/src/main/java/org/apache/parquet/schema/LogicalTypeAnnotation.java:874]: Sounds good!

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 5 above)
Comment (gszadovszky) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: I'll look into this just did not have time to work on this PR. Thanks a lot for reviewing. :)

File: parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java (hunk 3 above)
Comment (gszadovszky) [reply parquet-column/src/test/java/org/apache/parquet/schema/TestPrimitiveStringifier.java:322]: I'm happy to add tests for the edge cases like too short or too long inputs. Though, I would not implement additional validations because of performance issues. A `stringify` method would be invoked on each values; an additional check would highly impact performance even if it is only used from the tools and not really in production. A `Stringifier` is associated to the value at schema level which means it shall never happen that the value is invalid. That's why the `Stringifier` implementations do not validate the values.

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java (hunk 5 above)
Comment (gszadovszky) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:783]: The [`testRoundTripConversion`](https://github.com/apache/parquet-mr/blob/master/parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java#L144) I'm using in `testUUIDTypeWithParquetUUID` is actually stronger than the one you suggested: it checks for equality (in two phases) of the initial and the result avro schemas (and not only for compatibility). For `testUUIDType`, though it is a good idea to check the compatibility of the avro schemas.

Comment (shangxinli) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroSchemaConverter.java:786]: Sounds good
